### ISSUE TITLE
Too many compiler warnings

### ISSUE THREAD

**Issue by slindley** [links-lang/links#48]
Lots of compiler warnings are produced when Links is compiled. We should fix or suppress most of them.

**Comment by jstolarek** [comment]
@SimonJF, how actively are you working on this? I am currently familiarising myself with overall structure of the source code and I could fix warnings as I go. Just wouldn't want to get in your way with this.

**Comment by jstolarek** [comment]
I watched „What's new in OCamls 4.03” video from this year's ICFP and, according to it, warnings regarding in-place modification of strings will turn into errors at some point in the future.

**Comment by jstolarek** [comment]
While eliminating warnings I found this:

```ocaml
    method private with_rec_env recenv =
      {< rec_env = rec_env >}
```
The warning is that `recenv` is unused. Quite rightly! I assume the intended meaning was:
```ocaml
    method private with_rec_env rec_env =
      {< rec_env = rec_env >}
```
A good example why we should pay attention to warnings.

**Comment by jstolarek** [comment]
I started to silence the warnings but it's quite a lot of work and haven't even got halfway. My work is on [silence-warnings](https://github.com/links-lang/links/tree/silence-warnings) branch. It is based off the appserver2 branch, so it should be easy to merge it into sessions once the appserver branch is merged. In fact, this is just one large commit  so you can just cherry-pick (but please amend the commit message!)

```
git cherry-pick 75ced6c16062598c2e5f579a420472d8f1efb506
git commit --amend
```

@slindley said he'll be aiming to have this work finished before a release, so some things to keep in mind:

  * some functions have unused parameters. Remove those parameters rather than replace them with `_` - it often turns out that some value is computed and passed down through several calls only to be discarded
  * when dealing with non-exhaustive pattern matches try to enumerate all the possible values rather than add a catch-all pattern. This is a much better thing to do in case a data type gets new constructors in the future.
  * sometimes there are lambdas that do a pattern-match on some particular constructor and this generate a warning about non-exhaustive patterns. [Here's an example how I decide to deal with that](https://github.com/links-lang/links/commit/75ced6c16062598c2e5f579a420472d8f1efb506#diff-cec70da104fac15d671d881baf29e1c0R949). And [here is another](https://github.com/links-lang/links/commit/75ced6c16062598c2e5f579a420472d8f1efb506#diff-cec70da104fac15d671d881baf29e1c0R293)
  * You will sometimes see warnings about implicit elimination of optional arguments. [Here's my solution](https://github.com/links-lang/links/commit/75ced6c16062598c2e5f579a420472d8f1efb506#diff-cec70da104fac15d671d881baf29e1c0R666)

**Comment by slindley** [comment]
Now that the `appserver2` branch has been squash-merged into `sessions` the `silence-warnings` branch doesn't cleanly sync with it any more (lots of spurious merge conflicts). To sort this out, Garrett and I just created a `silence-warnings2` branch which is in sync with `sessions` and includes @jstolarek's one large commit. This is where further work on silencing warning will be carried out - in the next day or two.

**Comment by SimonJF** [comment]
I'm going through a couple of these; one thing I'm finding a bit is the "unused module" warning. This is generally due to currently-unused instances exported by deriving, which may (or may not) end up being used.

These are benign. I'm tempted to suggest suppressing the unused module warning; do people have opinions on this?

**Comment by SimonJF** [comment]
The CI seems to be failing on this branch since it can't find String.uncapitalize_ascii. This is because uncapitalize_ascii was introduced (and uncapitalize deprecated) in OCaml 4.03.0, whereas the seems to be running on 4.02.x.

Is there a way to bump up the compiler version on the CI, for example getting it to work in a different switch?

**Comment by jamescheney** [comment]
As long as whatever configuration needs ot be done can be done via a
command line or two, I think it would suffice to add those lines to the
"before_install" part of .travis.yml

https://github.com/links-lang/links/blob/sessions/.travis.yml

But I haven't tried that myself yet.
